File: hdf-generators/test/ruby-escape.test.ts
import { describe, it, expect } from 'vitest';
import { escapeQuotes } from '../src/ruby-escape.js';

describe('escapeQuotes', () => {
  it('wraps plain text in single quotes', () => {
    expect(escapeQuotes('hello world')).toBe("'hello world'");
  });

  it('wraps empty string in single quotes', () => {
    expect(escapeQuotes('')).toBe("''");
  });

  it('uses double quotes when string contains single quotes', () => {
    // In Ruby double-quoted strings, single quotes don't need escaping
    expect(escapeQuotes("it's a test")).toBe("\"it's a test\"");
  });

  it('uses single quotes when string contains double quotes', () => {
    // In Ruby single-quoted strings, double quotes don't need escaping
    expect(escapeQuotes('say "hello"')).toBe("'say \"hello\"'");
  });

  it('uses %q() when string contains both quote types', () => {
    expect(escapeQuotes(`it's a "test"`)).toBe(`%q(it's a "test")`);
  });

  it('escapes backslashes before closing parens in %q mode', () => {
    // \) → \\) so Ruby sees literal backslash, then ) closes %q() delimiter
    expect(escapeQuotes(`it's a "test" with \\)`)).toBe(`%q(it's a "test" with \\\\))`);
  });

  it('escapes backslashes in single-quoted strings', () => {
    expect(escapeQuotes('path\\to\\file')).toBe("'path\\\\to\\\\file'");
  });

  it('escapes backslashes in double-quoted strings', () => {
    // Double-quoted: backslashes escaped, single quotes pass through
    expect(escapeQuotes("it's a path\\to\\file")).toBe("\"it's a path\\\\to\\\\file\"");
  });

  it('handles multiline content', () => {
    const input = 'line one\nline two';
    const result = escapeQuotes(input);
    expect(result).toContain('line one');
    expect(result).toContain('line two');
  });

  it('handles strings with only backslashes', () => {
    expect(escapeQuotes('\\')).toBe("'\\\\'");
  });
});

File: hdf-generators/go/ruby_escape_test.go
package generators

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestEscapeQuotes_PlainText(t *testing.T) {
	assert.Equal(t, "'hello world'", EscapeQuotes("hello world"))
}

func TestEscapeQuotes_EmptyString(t *testing.T) {
	assert.Equal(t, "''", EscapeQuotes(""))
}

func TestEscapeQuotes_SingleQuotes(t *testing.T) {
	// String with single quotes → double-quoted
	assert.Equal(t, `"it's a test"`, EscapeQuotes("it's a test"))
}

func TestEscapeQuotes_DoubleQuotes(t *testing.T) {
	// String with double quotes → single-quoted
	assert.Equal(t, `'say "hello"'`, EscapeQuotes(`say "hello"`))
}

func TestEscapeQuotes_BothQuoteTypes(t *testing.T) {
	assert.Equal(t, `%q(it's a "test")`, EscapeQuotes(`it's a "test"`))
}

func TestEscapeQuotes_BackslashesInSingleQuoted(t *testing.T) {
	assert.Equal(t, `'path\\to\\file'`, EscapeQuotes(`path\to\file`))
}

func TestEscapeQuotes_BackslashesInDoubleQuoted(t *testing.T) {
	assert.Equal(t, `"it's a path\\to\\file"`, EscapeQuotes(`it's a path\to\file`))
}

func TestEscapeQuotes_BackslashBeforeParenInPercentQ(t *testing.T) {
	// \) → \\) in %q mode so Ruby sees literal backslash, then ) closes %q()
	assert.Equal(t, `%q(it's a "test" with \\))`, EscapeQuotes(`it's a "test" with \)`))
}

func TestEscapeQuotes_OnlyBackslashes(t *testing.T) {
	assert.Equal(t, `'\\'`, EscapeQuotes(`\`))
}

func TestEscapeQuotes_MultilineContent(t *testing.T) {
	result := EscapeQuotes("line one\nline two")
	assert.Contains(t, result, "line one")
	assert.Contains(t, result, "line two")
}
